### PR TITLE
Allow UTF-8 comments in program code

### PR DESCRIPTION
**Pull request by ahartikainen**
Allow UTF-8 in comments for `model_code` with `'utf-8'` on line 116 in api.py: `model_code_bytes = model_code.encode('utf-8')`.

#### Summary:

Use UTF-8 for characters in program code in `model_code_bytes = model_code.encode('utf-8')`. This allows utf-8 in comments.

#### Side Effects:

In python 2, raised error message is encoded with `.encode('ascii', 'replace')`

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: GPLv3 (http://opensource.org/licenses/GPL-3.0)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by bob-carpenter** [comment]
UTF-8 should be fine in comments without any munging on the Python side.  Once things get to the stanc parser in Stan, it just swallows comments as part of whitespace.

**Comment by ariddell** [comment]
Good point. I was worried about interactions with Cython but perhaps that's not going to be a problem. We should try just utf8 in that case.

**Comment by ariddell** [comment]
OK. We need two tests. (looking at @bob-carpenter 's comment on #223)

1. non-ascii text in comments doesn't cause any problems.
2. non-ascii text as, say, a parameter name in Stan program code does, in fact, raise the expected error/exception.

**Comment by bob-carpenter** [comment]
There's no test in Stan specifically for non-ASCII characters, it's just that identifiers are defined with [a-zA-Z0-9_] or something like that. So they shouldn't parse as Stan programs with non-ASCII variable names.

**Comment by ahartikainen** [comment]
I used unicode tag in tests due to python2: `u'text åäö`. 

I am still not satisfied with the error that comes from parameter being utf-8 char. It should be the same as in python 3 but could not fix it even after changing `.decode('ascii')` to `.decode('ascii', 'replace')` in `_api.pyx`.

Original error with `model_code = u'parameters {real ö;\n} model {ö ~ normal(0,1);}'` in python 2.7:

        ---------------------------------------------------------------------------
    UnicodeDecodeError                        Traceback (most recent call last)
    <ipython-input-10-ed2462a4f4e5> in <module>()
    ----> 1 model_4 = pystan.stanc(model_code=model_code4)

    /Users/ahartikainen/miniconda2/envs/tmp2/lib/python2.7/site-packages/pystan/api.pyc in stanc(file, charset, model_code, model_name, verbose, obfuscate_model_name)
        124     model_name_bytes = model_name.encode('ascii')
        125 
    --> 126     result = pystan._api.stanc(model_code_bytes, model_name_bytes)
        127     if result['status'] == -1:  # EXCEPTION_RC is -1
        128         error_msg = "Failed to parse Stan model '{}'. Error message:\n{}".format(model_name, result['msg'])

    /Users/ahartikainen/miniconda2/envs/tmp2/lib/python2.7/site-packages/pystan/_api.pyx in pystan._api.stanc (pystan/_api.cpp:996)()
         12     c_stanc(model_stancode, model_name, result)
         13     return {'status': result.status,
    ---> 14             'msg': result.msg.decode('ascii'),
         15             'model_cppname': result.model_cppname.decode('ascii'),
         16             'cppcode': result.cppcode.decode('ascii')}

    UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 97: ordinal not in range(128)

With python 3 the raised error is:

    ---------------------------------------------------------------------------
    ValueError                                Traceback (most recent call last)
    <ipython-input-12-61706e6e6582> in <module>()
    ----> 1 model_5 = pystan.stanc(model_code=model_code5)

    /Users/ahartikainen/miniconda2/envs/tmp/lib/python3.6/site-packages/pystan/api.py in stanc(file, charset, model_code, model_name, verbose, obfuscate_model_name)
        127     if result['status'] == -1:  # EXCEPTION_RC is -1
        128         error_msg = "Failed to parse Stan model '{}'. Error message:\n{}".format(model_name, result['msg'])
    --> 129         raise ValueError(error_msg)
        130     elif result['status'] == 0:  # SUCCESS_RC is 0
        131         logger.debug("Successfully parsed Stan model '{}'.".format(model_name))

    ValueError: Failed to parse Stan model 'anon_model_3e335373965de0fa43e0fa817541a9b4'. Error message:
    SYNTAX ERROR, MESSAGE(S) FROM PARSER:


    ERROR at line 1

      1:    parameters {real ö;
                            ^
      2:    } model {ö ~ normal(0,1);}

    PARSER EXPECTED: <identifier>

**Comment by ariddell** [comment]
For Python 2, can one just change:

``result.msg.decode('ascii')`` to ``result.msg.decode('utf8')``?

**Comment by ahartikainen** [comment]
I tested that too and it still fails. Not certain why. I will test again.

**Comment by ahartikainen** [comment]
Problem with PY2 was that raising errors work only with ascii.

File: pystan/api.py [hunk 1]
```diff
@@ -125,7 +125,11 @@ def stanc(file=None, charset='utf-8', model_code=None, model_name="anon_model",
 
     result = pystan._api.stanc(model_code_bytes, model_name_bytes)
     if result['status'] == -1:  # EXCEPTION_RC is -1
-        error_msg = "Failed to parse Stan model '{}'. Error message:\n{}".format(model_name, result['msg'])
+        msg = result['msg']
+        if PY2:
+            # fix problem with unicode in error message in PY2
+            msg = msg.encode('ascii', 'replace')
```
Comment (ariddell) [inline pystan/api.py]: This seems odd to me but it's been so long since I've programmed in Python 2.7 that I can't recall the compatibility tricks.

**Comment by ariddell** [comment]
LGTM. I'm happy to merge this unless you're interested in figuring out what's going on with the ValueError message encoding problem.

**Comment by ahartikainen** [comment]
Solution for PY2 to handle unicode in error messages is to use custom exception class. I don't see added value in that compared to `.encode('ascii', 'replace')`.

http://stackoverflow.com/questions/29096702/how-can-i-raise-an-exception-that-includes-a-unicode-string

You can merge.

**Comment by ariddell** [comment]
Interesting. I didn't realize that was the only way to do it.

**Pull request merged**